### PR TITLE
Save SoC info in results

### PR DESCRIPTION
**Pull request by d-uzlov**
Part of https://github.com/mlcommons/mobile_app_open/issues/508
Depends on https://github.com/mlcommons/mobile_app_open/pull/512

### CODE REVIEW

**Comment by anhappdev** [comment]
@d-uzlov Is this PR ready for review? If not, please mark it as draft.

**Comment by d-uzlov** [comment]
Anh, now this is ready for review.
Could you please also look into why Sonar build failed on this PR?
As you can see, there are no issues on iOS and Android CI builds, so there must be some difference in configs?

**Comment by anhappdev** [comment]
> Could you please also look into why Sonar build failed on this PR?
As you can see, there are no issues on iOS and Android CI builds, so there must be some difference in configs?

@d-uzlov 
I don't know yet why it failed, it has some additional argument for bazel build cmd, though. I will test later if those caused the issue.
https://github.com/mlcommons/mobile_app_open/blob/59dff25dca4d5a14fa5834d00ebe828b61b34c4b/tools/scanner/scan.mk#L28-L30

Can you provide the APK in the main issue so others can test if it works well on their devices?

**Comment by anhappdev** [comment]
The build is failed indeed by the additional flag: 
`sonar_bazel_build_args=--spawn_strategy=local --strategy=Genrule=local`
I tested it here: https://github.com/mlcommons/mobile_app_open/pull/528
Without it the build can success, but the code scan fails. It looks like there are many choices (https://blog.bazel.build/2019/06/19/list-strategy.html) so I need to try out which works.

**Comment by d-uzlov** [comment]
I believe that the issue is caused by duplicated files.
Cpuinfo is included in our workspace file and by tflite but with different names: `cpuinfo` and `org_pytorch_cpuinfo`.
Header file names are identical, so it's possible that compiler includes wrong file.
Also, `.c` files likely export symbols with the same name, but this likely doesn't trigger any errors.
I have no idea why sandboxed and local builds behave differently. I would expect both of them to fail.

We could just use cpuinfo from tflite but I would prefer not to because tflite uses old version of cpuinfo.
We could try to override tflite dependency on cpuinfo to use our version. I think it's the better solution.

**Review by anhappdev — CHANGES_REQUESTED** [review]

File: flutter_common/lib/data/environment/environment_info.dart
Comment (d-uzlov) [reply flutter_common/lib/data/environment/environment_info.dart:39]: I used cpuinfo as a single word here because it's the name of the library.
It's not "CPU Info", it's "data obtained from cpuinto lib".

Though, I don't have a strong preference here. If you think lib name shouldn't be used as a word here, I don't mind changing it.

File: flutter/lib/device_info.dart
Comment (d-uzlov) [reply flutter/lib/device_info.dart:16]: On iOS and Windows cpuinfo is able to provide CPU name.
On Android we often get `Unknown` instead of the CPU/SoC name when querying cpuinfo lib, so we try to gather some additional android-specific info.

File: flutter_common/lib/data/environment/environment_info.dart
Comment (anhappdev) [reply flutter_common/lib/data/environment/environment_info.dart:39]: We are writing Dart code here, so I think matching style `cpuInfo` would be better. If you want to say "data obtained from cpuinfo lib" then a comment is better.

File: flutter/lib/device_info.dart
Comment (anhappdev) [reply flutter/lib/device_info.dart:16]: Can we restructure or integrate that info into other field or class? It's quite strange that there is only one `androidInfo` and not other `iosInfo`, `windowsInfo`.

Comment (d-uzlov) [reply flutter/lib/device_info.dart:16]: I think it makes sense to put data that's only available on one OS into one field, and set this field to null when the OS is different.
If we mode data from `EnvAndroidInfo` into some other place, we will still have "this field is present on Android, null on other OS", except we will have several such fields.

File: flutter_common/lib/data/environment/environment_info.dart
Comment (d-uzlov) [reply flutter_common/lib/data/environment/environment_info.dart:39]: Would it be enough to rename this field to `libCpuinfo`?

File: flutter/lib/device_info.dart
Comment (anhappdev) [reply flutter/lib/device_info.dart:16]: This is mainly used to display `Device SoC/CPU name`? If yes, can we just merge / concatenate the `EnvAndroidInfo` into `EnvCpuinfo.socName`? Else there should be another `iosInfo` or `windowsInfo` because they certainly also have SoC info like model or manufacturer.

Comment (d-uzlov) [reply flutter/lib/device_info.dart:16]: `EnvAndroidInfo` does not contain actual cpu info.
It's some additional info that can be used to decode cpu info when it's missing.

>can we just merge / concatenate the EnvAndroidInfo into EnvCpuinfo

We can create fields:
`EnvCpuinfo.android_prop_soc_model`: nullable, only present on android, doesn't make sense on ios/windows.
`EnvCpuinfo.android_prop_soc_manufacturer`: nullable, only present on android, doesn't make sense on ios/windows.
`EnvCpuinfo.android_build_board`: nullable, only present on android, doesn't make sense on ios/windows.

I think it would be more confusing than putting these android-specific fields into a separate map.

>Else there should be another iosInfo or windowsInfo because they certainly also have SoC info like model or manufacturer

If we encounter any ios/windows devices that are missing `EnvCpuinfo.socName` field, we may have to create os-specific fields for them.
Right now theoretical `iosInfo` and `windowsInfo` fields are not needed because SoC model+manufacturer are contained in the `EnvCpuinfo.socName` field.

Comment (d-uzlov) [reply flutter/lib/device_info.dart:16]: > 1. Why is the `EnvCpuinfo.socName` not a field of `EnvSocInfo`?

Because `EnvCpuinfo.socName` is data from `cpuinfo` lib.

> 2\. Why is there only `EnvAndroidInfo` but not for iOS, macOS or Windows (as the original question)?

Because `EnvCpuinfo.socName` is different on different operating systems.
On iOS and Windows it always contain valid data.
On Android there is  50/50 chance that `EnvCpuinfo.socName` is valid, so we have to add os-specific fields to make CPU detection more consistent.

File: flutter_common/lib/data/environment/environment_info.dart
Comment (anhappdev) [reply flutter_common/lib/data/environment/environment_info.dart:39]: Either `cpuInfo` or `cpuinfo` I understand them both as  "info about cpu". I assume others may also understand so. So in this case it should have a different name to avoid confusion.

File: flutter/lib/device_info.dart
Comment (d-uzlov) [reply flutter/lib/device_info.dart:16]: > The structure is inconsistent.

The structure is: `environmentInfo map` -> `socInfo map` -> `data source type` -> `actual data`.

File: flutter_common/lib/data/environment/environment_info.dart
Comment (d-uzlov) [reply flutter_common/lib/data/environment/environment_info.dart:39]: > Would it be enough to rename this field to `libCpuinfo`?

I propose `libCpuinfo`.
It would be clear that this is not some general info but an info from a specific source.

We could use `libCpuInfo` if you insist on using proper English words.
I would prefer to keep the `cpuinfo` reference in the name but I would be OK with changing it.

File: flutter/lib/device_info.dart
Comment (anhappdev) [reply flutter/lib/device_info.dart:16]: I suggest using this structure to have the data unified.
```dart
enum EnvSocInfoDataSourceEnum { cpuinfo, getprop, board, ... }
class EnvSocInfo {
  final String socName;
  final EnvSocInfoDataSourceEnum dataSource;
}
```

Comment (d-uzlov) [reply flutter/lib/device_info.dart:16]: What would be the algorithm to ensure that `socName` is valid and consistent?

iOS and Windows are OK since on these systems we have single reliable source of data.

On Android currently we use the following algorithm:
If board has a known value, use board.
If getprop returned valid data, use it.
Else fall back to cpuinfo.

The problem here is that `board` value is mapped to a proper name via a database that may be temporary outdated.
So we can't decode board in current version of the app but we will be able to do it in later versions.

This is why I kept all of the source data. If we have all the data, we may potentially be able to show something readable even if we encounter unknown board value.

Comment (anhappdev) [reply flutter/lib/device_info.dart:16]: The main question is what to do with `socName`. If it's only used for UI display, just make `socName` nullable and display what it has.
If you want to keep all the data as detailed as possible, then I suggest using this structure:
```dart
socInfo: EnvSocInfo(
  iosInfo: EnvSocInfoIos(
    socName: 'soc name',
    iosOnlyField: 'ios only data'
  ),
  windowsInfo: EnvSocInfoWindows(
    socName: 'soc name'
  ),
  androidInfo: EnvSocInfoAndroid(
    socName: 'soc name'
    propSocModel: 'soc model',
    propSocManufacturer: 'soc manufacturer',
    buildBoard: 'board'
  ),
),
```
The point is that all fields in the same level of hierarchy should have something in common. I don't think that the choice of underlying 3rd party libs should impact the data structure.

Comment (d-uzlov) [reply flutter/lib/device_info.dart]: I changed environment info data layout to have device-specific fields for each OS.
Except It made more sense to me to also separate model name and stuff.

**Comment by d-uzlov** [comment]
@anhappdev maybe we should disable quality gate for code coverage in Sonar?
We don't have any tests for native code, so it's natural to have 0% coverage.
We can change it later if we decide to spend time to set up unit testing for C++ code.

**Comment by anhappdev** [comment]
> @anhappdev maybe we should disable quality gate for code coverage in Sonar?
> We don't have any tests for native code, so it's natural to have 0% coverage.
> We can change it later if we decide to spend time to set up unit testing for C++ code.

@morphine00 Bruno, can you help us with this?

**Comment by morphine00** [comment]
Ok, I've copied the default rule set but disabled the coverage rule.

I did see this message, let me know if anything needs to be done about this:

`This Quality Gate sets conditions on new code but you do not have a new code definition set. If you do not set a new code definition, the Quality Gate will only appear on your pull requests and short-lived branches, but not on your main branch or other long-lived branches. To enable it for these cases, set your new code definition.`

**Comment by anhappdev** [comment]
@morphine00 Thanks for your help. Regarding the new code definition warning, this PR https://github.com/mlcommons/mobile_app_open/pull/583 should fix it.

File: flutter/lib/device_info.dart
Comment (anhappdev) [inline flutter/lib/device_info.dart:43]: Typo: `makeWIndows` -> `makeWindows`

File: flutter/lib/device_info.dart
Comment (anhappdev) [inline flutter/lib/device_info.dart:101]: `type` and `name` look redundant here? I would just define 2 fields `key` and `value` for `EnvAndroidProp`.
I guess you want to use the field `type` as a filter? Then it should be a `enum`, since it should have a small set of values.

**Review by anhappdev — CHANGES_REQUESTED** [review]

File: flutter/android/app/build.gradle
Comment (d-uzlov) [reply flutter/android/app/build.gradle:29]: It became required after I merged the commit with flutter version update.
Android build on CI failed with error:
```
Dependency 'androidx.core:core:1.9.0' requires 'compileSdkVersion' to be set to 33 or higher.
Compilation target for module ':app' is 'android-31'

Dependency 'androidx.core:core-ktx:1.9.0' requires 'compileSdkVersion' to be set to 33 or higher.
Compilation target for module ':app' is 'android-31'

Dependency 'androidx.annotation:annotation-experimental:1.3.0' requires 'compileSdkVersion' to be set to 33 or higher.
Compilation target for module ':app' is 'android-31'
```
https://console.cloud.google.com/cloud-build/builds/b9f6c2a5-e5b8-4cc1-b37f-b6f0704fb567;step=5?project=mobile-app-build-290400

Comment (d-uzlov) [reply flutter/android/app/build.gradle:29]: > Also, the change `"platforms;android-33"` in `flutter/android/docker/Dockerfile`?

By the way, it doesn't really matter. Flutter installs all required android tools and SDKs automatically if it has write access to android home folder. The only thing that changes is we can save a bit of time if we already have it installed.

File: flutter/documentation/result-spec.md
Comment (d-uzlov) [reply flutter/documentation/result-spec.md:117]: > Would `platform` instead of `device_type` more suitable here?

Yeah, maybe. Sounds good to me, I'll change it.

> What would be all possible values?

I was thinking that maybe it would be beneficial to have a separate field for windows_arm devices, if they need some special handling (for example, full_cpu_name field may not work on ARM, and we would need to do something like `board` on android).
Maybe a separate type for windows laptops, since thay can possibly have a defined manufacturer and model.

File: flutter/lib/device_info.dart
Comment (d-uzlov) [reply flutter/lib/device_info.dart:101]: > I guess you want to use the field `type` as a filter? Then it should be a `enum`, since it should have a small set of values.

Yeah.

`name` field is useless here, As in: we will not show it, we will not use it for any logic.
But when we have working online result database we could use it to see which values are present on which devices.
Though, it should be noted that we are not saving the whole `getprop` output, we are only saving specified values, so will still need to manually determine list of values we are interested in, so the app saves them in result json.

File: flutter/android/app/build.gradle
Comment (anhappdev) [reply flutter/android/app/build.gradle:29]: The `share_plus` package resolved to `6.0.1` in `flutter/pubspec.lock`, which leads to the crash. We can fix the version to `share_plus: 6.0.0` in `flutter/pubspec.yaml` in this PR. And maybe create another PR to update the `compileSdkVersion` if needed.

Comment (d-uzlov) [reply flutter/android/app/build.gradle]: done

File: flutter/documentation/result-spec.md
Comment (d-uzlov) [reply flutter/documentation/result-spec.md]: done

File: flutter/lib/device_info.dart
Comment (d-uzlov) [reply flutter/lib/device_info.dart]: fixed

File: flutter/lib/device_info.dart
Comment (d-uzlov) [reply flutter/lib/device_info.dart]: done

**Review by anhappdev — APPROVED** [review]

**Pull request merged**
